src/interfaces/sampler_interface.jl

```diff
@@ -0,0 +1,125 @@
+#################
+# The sampler interface
+#
+# In this file, we define the interface for general samplers and specific
+# samplers for scattering processes.
+#################
+
+"""
+Abstract base type for general sampler.
+
+!!! note "Sampler interface"
+
+    Functions to be implemented
+
+    ```Julia
+    CustomSampler <: AbstractSampler{Float64}
+
+    Base.eltype  # type of the generated samples
+    QEDevents._weight
+    Base.size  # Return the dimensionality of the sampler (number of outgoing four-momenta per sample).
+    is_exact
+    QEDevents._rand!(rng::AbstractRNG, ::AbstractSampler, x::AbstractVector{T}) where {T}
+
+    ```
+    
+    optional
+
+    ```Julia
+    _post_processing
+    ```
+
+"""
+abstract type AbstractSampler <: AbstractComputationSetup end
+
+@inline Base.eltype(::AbstractSampler) = Float64
+
+"""
+Interface function, which asserts that the given `input` is valid.
+"""
+function _assert_valid_input(smplr::AbstractSampler, x::AbstractVecOrMat)
+    size(x, 1) == size(smplr, 1)[1] || error(
+        "Invalid input. The dimensionality of the input is $(size(x,1)) but it should be $(size(smplr,1)[1]).",
+    )
+    return eltype(x) == eltype(smplr) || error(
```

Comment:
Same: Instead of using the function `error("...")`, consider using `throw(InvalidInputError("..."))`.

Reply:
Implemented in 29e04d4dc84862ce3aabe9a8addce347bdc5f0a7